lib/logger/lib/logger/app.ex

```diff
@@ -21,41 +23,69 @@ defmodule Logger.App do
                     {otp_reports?, sasl_reports?, threshold}, :link],
                   [id: Logger.ErrorHandler, function: :watcher])]
 
+    config = Logger.Config.new()
     case Supervisor.start_link(children, options) do
-      {:ok, _} = ok ->
+      {:ok, sup} ->
         deleted = delete_error_logger_handler(otp_reports?, :error_logger_tty_h, [])
         deleted = delete_error_logger_handler(sasl_reports?, :sasl_report_tty_h, deleted)
         store_deleted_handlers(deleted)
-        ok
+        {:ok, sup, config}
       {:error, _} = error ->
         error
     end
   end
 
   @doc false
-  def stop(_) do
-    Application.get_env(:logger, :deleted_handlers)
-    |> Enum.each(&:error_logger.add_report_handler/1)
-
-    # We need to do this in another process as the Application
-    # Controller is currently blocked shutting down this app.
-    spawn_link(fn -> Logger.Config.clear_data end)
+  def stop(config) do
+    Logger.Config.delete(config)
+    try do
+      get_deleted_handlers()
+    rescue
+      # Quiet stop occured, handlers will be re-added by caller after stoppng
+      ArgumentError ->
+        :ok
+    else
+      deleted_handlers ->
+        Enum.each(deleted_handlers, &:error_logger.add_report_handler/1)
+        delete_deleted_handlers()
+    end
+  end
 
-    :ok
+  @doc false
+  def config_change(_changed, _new, _removed) do
+    Logger.Config.configure([])
   end
 
   @doc """
   Stops the application without sending messages to error logger.
   """
   def stop() do
-    set = Application.get_env(:logger, :deleted_handlers)
-    Application.put_env(:logger, :deleted_handlers, HashSet.new)
-    _ = Application.stop(:logger)
-    Enum.each(set, &:error_logger.add_report_handler/1)
+    try do
+      get_deleted_handlers()
+    rescue
+      ArgumentError ->
+        {:error, {:not_started, :logger}}
+    else
+      deleted_handlers ->
+        delete_deleted_handlers()
+        result = Application.stop(:logger)
+        Enum.each(deleted_handlers, &:error_logger.add_report_handler/1)
+        result
+    end
   end
 
   defp store_deleted_handlers(list) do
-    Application.put_env(:logger, :deleted_handlers, Enum.into(list, HashSet.new))
+    tab = :ets.new(@table, [:named_table, :public])
```

Comment:
Maybe we can share this table with the Logger.Config one? Possibly store this stuff in the Logger.Config itself?